File: tools/improve/tests/devstral-small-2/maudec/py/stack2.py
#
#	<comment from="devstral-small-2">
#	I'll simplify and improve the efficiency of the Python code while preserving the semantics and keeping the original function signatures.
#	</comment>
#

import enum

class StackNat(enum.Enum):
    Mt = 0  # Empty stack
    Push = 1  # int, StackNat
    Err = 2  # Error


def pop(a1: StackNat) -> StackNat:
    if a1 == StackNat.Mt:
        return StackNat.Mt
    return a1.value[2]  # Assuming the structure is (Push, N, S)


def top(a1: StackNat) -> int:
    if a1 == StackNat.Push:
        return a1.value[1]  # Assuming the structure is (Push, N, S)
    raise ValueError("Cannot get top of empty stack")


def empty(S: StackNat) -> bool:
    return S == StackNat.Mt


def topIncr(a1: StackNat) -> StackNat:
    if a1 == StackNat.Mt:
        return StackNat.Mt
    return StackNat.Push, a1.value[1] + 1, a1.value[2]


def incr(a1: StackNat) -> StackNat:
    if a1 == StackNat.Mt:
        return StackNat.Mt
    return StackNat.Push, a1.value[1] + 1, incr(a1.value[2])


def merge(S: StackNat) -> StackNat:
    if S == StackNat.Push and S.value[2] == StackNat.Push:
        return StackNat.Push, S.value[1] + S.value[2].value[1], S.value[2].value[2]
    return S

File: tools/improve/tests/devstral-small-2/maudec/rust/stack2.rs
//
//	<comment from="devstral-small-2">
//	I'll simplify and improve the efficiency of the Rust code while preserving the semantics and keeping the original function signatures.
//	</comment>
//

#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum Stack<i32> {
    Mt,
    Push(i32, Box<Stack<i32>>),
    Err,
}

fn pop(a1: &Stack<i32>) -> Stack<i32> {
    match a1 {
        Stack::Mt => Stack::Mt,
        Stack::Push(_, s) => *s.clone(),
        _ => panic!("case not covered"),
    }
}

fn top(a1: &Stack<i32>) -> i32 {
    match a1 {
        Stack::Push(e, _) => *e,
        _ => panic!("case not covered"),
    }
}

fn empty(s: &Stack<i32>) -> bool {
    matches!(s, Stack::Mt)
}

fn topIncr(a1: &Stack<i32>) -> Stack<i32> {
    match a1 {
        Stack::Mt => Stack::Mt,
        Stack::Push(n, s) => Stack::Push(n + 1, s.clone()),
        _ => panic!("case not covered"),
    }
}

fn incr(a1: &Stack<i32>) -> Stack<i32> {
    match a1 {
        Stack::Mt => Stack::Mt,
        Stack::Push(n, s) => Stack::Push(n + 1, Box::new(incr(s))),
        _ => panic!("case not covered"),
    }
}

fn merge(s: &Stack<i32>) -> Stack<i32> {
    match s {
        Stack::Push(n, paux1) if matches!(paux1, Stack::Push(..)) => {
            if let Stack::Push(x, rest) = &**paux1 {
                Stack::Push(n + x, rest.clone())
            } else {
                panic!("unchecked variant access")
            }
        }
        s => s.clone(),
    }
}